Sources/NIO/ByteBuffer-core.swift

```diff
@@ -800,6 +804,53 @@ extension ByteBuffer {
     }
 }
 
+extension ByteBuffer {
+    /// Errors thrown when calling `copyBytes`.
+    public enum CopyBytesError: Error, Equatable {
+        /// The length of the bytes to copy was negative.
+        case negativeLength
+
+        /// The bytes to copy are not readable.
+        case unreadableSourceBytes
+    }
+
+    /// Copies `length` `bytes` starting at the `fromIndex` to `toIndex`. Does not move the writer index.
+    ///
+    /// - Note: Overlapping ranges, for example `copyBytes(at: 1, to: 2, length: 5)` are allowed.
+    /// - Precondition: The range represented by `fromIndex` and `length` must be readable bytes,
+    ///     that is: `fromIndex >= readerIndex` and `fromIndex + length <= writerIndex`.
+    /// - Parameter fromIndex: The index of the first byte to copy.
+    /// - Parameter toIndex: The index into to which the first byte will be copied.
+    /// - Parameter length: The number of bytes which should be copied.
+    @discardableResult
+    @inlinable
+    public mutating func copyBytes(at fromIndex: Int, to toIndex: Int, length: Int) throws -> Int {
+        guard length > 0 else {
+            if length == 0 {
+                return 0
+            } else {
+                throw CopyBytesError.negativeLength
+            }
+        }
+        guard self.readerIndex <= fromIndex && fromIndex + length <= self.writerIndex else {
+            throw CopyBytesError.unreadableSourceBytes
+        }
+
+        let newEndIndex = max(self._writerIndex, _toIndex(toIndex + length))
```

Comment:
We may as well do this math inside the `if` block and save ourselves some computation.